cmd/juju/cloud/add_test.go

```diff
@@ -7,172 +7,256 @@ import (
+//       new-york:
+//         endpoint: http://newyork/1.0
+//   garage-maas:
+//     type: mass
+//     auth-types: [oauth]
+//     endpoint: http://garagemaas
+// `[1:]
+// 	err = ioutil.WriteFile(sourceFile, []byte(source), 0600)
+// 	c.Assert(err, jc.ErrorIsNil)
+// 	return sourceFile
+// }
+
 func (s *addSuite) TestAddBadFilename(c *gc.C) {
-	addCmd := cloud.NewAddCloudCommand()
+	fake := newFakeCloudQueryer()
+	badFileErr := errors.New("")
+	fake.Call("ParseCloudMetadataFile", "somefile.yaml").Returns(map[string]cloudfile.Cloud{}, badFileErr)
+
+	addCmd := cloud.NewAddCloudCommand(fake)
 	_, err := testing.RunCommand(c, addCmd, "cloud", "somefile.yaml")
-	c.Assert(err, gc.ErrorMatches, "open somefile.yaml: .*")
+	c.Check(err, gc.Equals, badFileErr)
 }
 
 func (s *addSuite) TestAddBadCloudName(c *gc.C) {
-	sourceFile := s.createTestCloudData(c)
-	addCmd := cloud.NewAddCloudCommand()
-	_, err := testing.RunCommand(c, addCmd, "cloud", sourceFile)
+	fake := newFakeCloudQueryer()
+	fake.Call("ParseCloudMetadataFile", "testFile").Returns(map[string]cloudfile.Cloud{}, nil)
+
+	addCmd := cloud.NewAddCloudCommand(fake)
+	_, err := testing.RunCommand(c, addCmd, "cloud", "testFile")
 	c.Assert(err, gc.ErrorMatches, `cloud "cloud" not found in file .*`)
 }
 
 func (s *addSuite) TestAddExisting(c *gc.C) {
-	sourceFile := s.createTestCloudData(c)
-	_, err := testing.RunCommand(c, cloud.NewAddCloudCommand(), "homestack", sourceFile)
+	fake := newFakeCloudQueryer()
+	fake.Call("ParseCloudMetadataFile", "fake.yaml").Returns(homestackMetadata(), nil)
+	fake.Call("PublicCloudMetadata", []string(nil)).Returns(map[string]cloudfile.Cloud{}, false, nil)
+	fake.Call("PersonalCloudMetadata").Returns(homestackMetadata(), nil)
+
+	_, err := testing.RunCommand(c, cloud.NewAddCloudCommand(fake), "homestack", "fake.yaml")
 	c.Assert(err, gc.ErrorMatches, `"homestack" already exists; use --replace to replace this existing cloud`)
 }
 
 func (s *addSuite) TestAddExistingReplace(c *gc.C) {
-	sourceFile := s.createTestCloudData(c)
-	_, err := testing.RunCommand(c, cloud.NewAddCloudCommand(), "homestack", sourceFile, "--replace")
+	fake := newFakeCloudQueryer()
+	fake.Call("ParseCloudMetadataFile", "fake.yaml").Returns(homestackMetadata(), nil)
+	fake.Call("PersonalCloudMetadata").Returns(homestackMetadata(), nil)
+	numCallsToWrite := fake.Call("WritePersonalCloudMetadata", homestackMetadata()).Returns(nil)
+
+	_, err := testing.RunCommand(c, cloud.NewAddCloudCommand(fake), "homestack", "fake.yaml", "--replace")
 	c.Assert(err, jc.ErrorIsNil)
-	data, err := ioutil.ReadFile(osenv.JujuXDGDataHomePath("clouds.yaml"))
-	c.Assert(string(data), gc.Equals, `
-clouds:
-  homestack:
-    type: openstack
-    auth-types: [userpass, access-key]
-    endpoint: http://homestack
-    regions:
-      london:
-        endpoint: http://london/1.0
-      new-york:
-        endpoint: http://newyork/1.0
-`[1:])
+
+	c.Check(numCallsToWrite(), gc.Equals, 1)
```

Comment:
> So, in this case, what the test is actually testing is that the old data is replaced by the new data in Write. [...] The only thing it is explicitly says it is checking is that write was called once. It doesn't have a line that says "check if what was written was correct".

I disagree with this premise. The test is correct in that it is only checking that it's gluing the bits together correctly, i.e.: correct function call, args, and results. This is what the tests is testing. See my prior comment about the role of unit/functional tests.

> It also means that when the test fails, it'll give a really terrible failure message. Instead of a very obvious 'expected to write "foo: bar", instead wrote "baz: bat"', you'll get "expected 1, got 0".

We were both wrong; I forgot, that function call will never be made. If you don't put in the correct arguments, you'll get a panic[1] with the errant method call listed. I think maybe we can clean this up a little, but that's a PR to the mock code.

The check for whether or not the mocked function is called is to guard against multiple calls, or an unexpected call to another registered mock call.

[1]
```
PANIC: add_test.go:428: addSuite.TestInteractiveExistingNameNoOverride

[LOG] 0:00.000 DEBUG  Call: PublicCloudMetadata([[]])
[LOG] 0:00.000 DEBUG  Results: [map[] false <nil>]
[LOG] 0:00.000 DEBUG  Call: PersonalCloudMetadata([])
[LOG] 0:00.000 DEBUG  Results: [map[homestack:{openstack  [userpass access-key] http://homestack   [{london http://london/1.0  }] map[] map[]}] <nil>]
[LOG] 0:00.000 DEBUG  Call: ParseOneCloud([[101 110 100 112 111 105 110 116 58 32 49 57 50 46 49 54 56 46 49 46 54 10]])
[LOG] 0:00.000 DEBUG  Results: [{manual  [] 192.168.1.6   [] map[] map[]} <nil>]
[LOG] 0:00.000 DEBUG  Call: PersonalCloudMetadata([])
[LOG] 0:00.000 DEBUG  Results: [map[homestack:{openstack  [userpass access-key] http://homestack   [{london http://london/1.0  }] map[] map[]}] <nil>]
[LOG] 0:00.000 DEBUG  Call: WritePersonalCloudMetadata([map[homestack:{openstack  [userpass access-key] http://homestack   [{london http://london/1.0  }] map[] map[]} homestack2:{manual  [] 192.168.1.6   [] map[] map[]}]])
[LOG] 0:00.000 DEBUG  Results: []
... Panic: runtime error: index out of range (PC=0x45D80C)

/gnu/store/6i15c20ggyix8q4sy5z63izfwfkz24q9-go-1.7.3/src/runtime/panic.go:458
  in gopanic
/gnu/store/6i15c20ggyix8q4sy5z63izfwfkz24q9-go-1.7.3/src/runtime/panic.go:27
  in panicindex
add_test.go:56
  in fakeCloudMetadataStore.WritePersonalCloudMetadata
add.go:303
  in addCloud
add.go:165
  in addCloudCommand.runInteractive
add.go:112
  in addCloudCommand.Run
add_test.go:459
  in addSuite.TestInteractiveExistingNameNoOverride
/gnu/store/6i15c20ggyix8q4sy5z63izfwfkz24q9-go-1.7.3/src/reflect/value.go:302
  in Value.Call
/gnu/store/6i15c20ggyix8q4sy5z63izfwfkz24q9-go-1.7.3/src/runtime/asm_amd64.s:2086
  in goexit
OOPS: 15 passed, 1 PANICKED
--- FAIL: TestPackage (0.05s)
FAIL
FAIL	github.com/juju/juju/cmd/juju/cloud	0.095s
```